File: Reflectance_S.py
#%%
import numpy as np
from math import *

#%%
# (a)set wavelength and angle of incident light

# set wavelength
lambda_0 = 500
# convert angle to radians and change convention
angle = 0 * (pi / 180)


#%%
# (b)set structural parameters of multilayer

# set the number of period in multilayer
Num = 5
# initialize the thickness for each layer
z = np.zeros((2 * Num + 2, 1))
# initialize the optical index for each layer
F = np.zeros((2 * Num + 2, 1))
# set the thickness of incident and exit medium
z[0] = 10
z[-1] = 10
# set the optical index of incident and exit medium
F[0] = 1
F[-1] = 1
# calculate number of layers in ML
N = len(z)
# set the index of each material in multilayer
nH = 2.5
nL = 1.25
# set optical thickness of layer to 1/4 wavelength
dH = (1 / 4) * lambda_0 / np.real(nH)
# set optical thickness of layer to 1/4 wavelength
dL = (1 / 4) * lambda_0 / np.real(nL)

# assign thickness and index to each layer in multilayer
for m in range(1, Num + 1):
    z[2 * m - 1] = dH
    # low index material
    F[2 * m - 1] = nH
    z[2 * m] = dL
    # high index material
    F[2 * m] = nL
#%%
# (c)initialize calculating values
theta = np.zeros((N, 1))
theta[0] = angle
#%%
# (d)calculating r and t of multilayer
M = np.array([[1, 0], [0, 1]])
# changing interface/layers
for k in range(2, N):
    # determine refractive index of upper layer
    n1 = F[k - 1]
    # determine refractive index of this layer
    n2 = F[k - 2]
    # incident angle of this layer
    theta[k - 1] = asin(sin(theta[k - 1]) * n2 / n1)
    # pre-determine repeating constants (to save time)
    K = 2 * pi * cos(theta[k - 1]) / lambda_0
    # 有效位相厚度
    Q = K * n1 * z[k - 1]
    # optical admittance (for now set to S-polarized)
    p = n1 * cos(theta[k - 1])
    # a,b,c,d: component of 2x2 matrix representing each ML interface
    a = cos(Q)
    b = 1j * sin(Q) / p
    c = 1j * p * sin(Q)
    # d = cos(Q)
    # M is the product of m1*m2*...
    M = M @ np.array([[a, b], [c, a]], dtype=object)

# angle in emergent medium
theta[N - 1] = asin(sin(theta[0]) * F[0] / F[N - 1])
# optical admittance incident and emergent medium
p1 = F[0] * cos(theta[0])
ps = F[N - 1] * cos(theta[N - 1])

# S = M * [1; ps]
S = M @ np.array([[1], [ps]], dtype=object)
B = S[0, 0]
C = S[1, 0]
# 等效光学导纳
Y = C / B
# reflection coefficient
r = (p1 - Y) / (p1 + Y)
# reflectivity
Refl = abs(r) ** 2
print("Refl = ", Refl[0])

t = 2 * p1 / (B * p1 + C)
Trans = 4 * np.real(p1) * np.real(ps) / (abs(p1 * B + C) ** 2)
print("Trans = ", Trans[0])

# %%

File: Reflectance_S.m
%Program Reflectance of optical thin film----for S polarization (by chengyou lin-2013.9.16)
clear all; clc %saves grief by clearing previous variables and clearing screen

%(a)set wavelength and angle of incident light(start)

lambda = 500; %set wavelength

angle = 0 * (pi / 180); %convert angle to radians and change convention
%(a)set wavelength and angle of incident light(over)

%(b)set structural parameters of multilayer(start)
Num = 5; %set the number of period in multilayer
z = zeros(2 * Num + 2, 1); %initialize the thickness for each layer
F = zeros(2 * Num + 2, 1); %initialize the  optical index for each layer

z(1) = 10; z(2 * Num + 2) = 10; %set the thickness of incident and exit medium
F(1) = 1; F(2 * Num + 2) = 1; %set the optical index of incident and exit medium

N = length(z); %calculate number of layers in ML

nH = 2.5; nL = 1.25; % set the index of each material in multilayer

dH = (1/4) * lambda / real(nH); %set optical thickness of layer to 1/4 wavelength
dL = (1/4) * lambda / real(nL); % %set optical thickness of layer to 1/4 wavelength

%assign thickness and index to each layer in multilayer
for m = 1:Num
    z(2 * m) = dH;
    F(2 * m) = nH; %low index material
    z(2 * m + 1) = dL;
    F(2 * m + 1) = nL; %high index material
end

%assign thickness and index to each layer in multilayer
%(b)set structural parameters of multilayer(over)

%{
%(b)set structural parameters of multilayer(start)
na = 1; ng = 1; nH = 2.5; nL = 1.25;
da = 10; dg = 10;
dH = (1/4) * lamda0 / real(nH); %set optical thickness of layer to 1/4 wavelength
dL = (1/4) * lamda0 / real(nL); % %set optical thickness of layer to 1/4 wavelength
z = [da; dH; dL; dH; dL; dH; dL; dH; dL; dH; dL; dg]; F = [na; nH; nL; nH; nL; nH; nL; nH; nL; nH; nL; ng];
N = length(z); %calculate number of layers in ML

%(b)set structural parameters of multilayer(over)

%}

%{
%(b)set structural parameters of multilayer(start)
fid1 = '001.txt'; [Fn, Fk, z] = textread(fid1, '%f %f %f');
F = Fn - Fk * 1i;
N = length(z); %calculate number of layers in ML

%(b)set structural parameters of multilayer(over)

%}

%(c)initialize calculating values (start)
theta = zeros(N, 1); theta(1) = angle;
%(c)initialize calculating values (over)

%(d)calculating r and t of multilayer(start)
% for m = 1:length(lamda) %changing Lambda

M = [1, 0; 0, 1];

for k = 2:N - 1 %changing interface/layers
    n1 = F(k); %determine refractive index of upper layer
    n2 = F(k - 1); %determine refractive index of this layer
    theta(k) = asin(sin(theta(k - 1)) * n2 / n1); %incident angle of this layer

    K = 2 * pi * cos(theta(k)) ./ lambda; %pre-determine repeating constants (to save time)
    Q = K * n1 * z(k); %有效位相厚度
    p = n1 * cos(theta(k)); %optical admittance (for now set to S-polarized)

    a = cos(Q); %a,b,c,d: component of 2x2 matrix representing each ML interface
    b = 1i * sin(Q) / p;
    c = 1i * p * sin(Q);
    %d = cos(Q);
    M = M * [a, b; c, a]; %M is the product of m1*m2*...

end

theta(N) = asin(sin(theta(1)) * F(1) / F(N)); % angle in emergent medium
p1 = F(1) * cos(theta(1)); ps = F(N) * cos(theta(N)); %optical admittance incident and emergent medium
S = M * [1; ps];
B = S(1, 1); C = S(2, 1);
Y = C / B; %等效光学导纳
r = (p1 - Y) / (p1 + Y); %reflection coefficient
Refl = abs(r)^2 %reflectivity

t = 2 * p1 / (B * p1 + C);
Trans = 4 * real(p1) * real(ps) / (abs(p1 * B + C)^2)
